CMake for reg2grb2

BrianCurtis-NOAA opened the pull request:
Add a CMakeLists.txt to be able to create reg2grb2.x for coupled-crow after the merge of CMake work from https://github.com/NOAA-EMC/global-workflow/pull/264

Linked issue:
#263 

Waiting on:
Pull #264

BrianCurtis-NOAA committed:
CMake for reg2grb2

JessicaMeixner-NOAA commented:
Maybe it's just me, but this seems it could be independent of  #264 and doesn't really need to wait for it?In addition, as this does not delete sorc/reg2grb2.fd/Makefile or update the build script sorc/build_reg2grb2.sh  and is just adding a file, couldn't this just go in now?

BrianCurtis-NOAA committed:
*Forgot -free
*Corrected link lib names

BrianCurtis-NOAA commented:
I did test the file manually putting it into Rahul's CMake work with feature/cmake, and it built the executable.
You are correct that the file can go in now without any issue.
Once the cmake work is brought over to feature/coupled-crow I can make the edits to CMakeLists.txt get it where it needs to be.

BrianCurtis-NOAA committed:
import statements should be on their own lines, pep8

KateFriedman-NOAA commented:
@BrianCurtis-NOAA Checking in...is this draft PR still needed? Thanks!

BrianCurtis-NOAA commented:
Has reg2grb2 been removed or already ported to cmake elsewhere? I believe I was waiting on the CMake build to be ported into coupled-crow.

BrianCurtis-NOAA committed:
Merge branch 'feature/coupled-crow' of https://github.com/NOAA-EMC/global-workflow into feature/coupled-crow

BrianCurtis-NOAA committed:
Merge branch 'feature/coupled-crow' into feature/reg2grb2_cmake

BrianCurtis-NOAA committed:
remove changes to rocoto_viewer

BrianCurtis-NOAA commented:
Yeah missed the cmake port completely. Let me get this added to the sorc CMakeLists.txt.
@WalterKolczynski-NOAA how would I test this is working once I get these last changes in?

BrianCurtis-NOAA committed:
Get CMake supported reg2grb2 into build

WalterKolczynski-NOAA commented:
As long as it builds and runs, it's fine as far as I'm concerned. I don't have any tests to confirm output for reg2grb2.

BrianCurtis-NOAA commented:
OK, I'll test on hera and switch this to ready when done.

BrianCurtis-NOAA commented:
Moving to ready.

BrianCurtis-NOAA commented:
Please hold for now while I check on something.

BrianCurtis-NOAA committed:
Add find_package in CMakeLists.txt for reg2grb2 additions

BrianCurtis-NOAA commented:
I left out the find package calls, my reg2grb2 builds OK, but the build on sorc fails with the `enkf_chgres_recenter_nc.fd` directory 
```
[ 19%] Building Fortran object regrid_nemsio.fd/CMakeFiles/regrid_nemsio.x.dir/netcdfio_interface.f90.o
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(5): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [MODULE_NCIO]
 use module_ncio
-----^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(53): error #6457: This derived type name has not been declared.   [DATASET]
 type(Dataset) :: indset
------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(54): error #6457: This derived type name has not been declared.   [DIMENSION]
 type(Dimension) :: ncdim
------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(319): error #6457: This derived type name has not been declared.   [DATASET]
 type(Dataset) :: refdset
------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(68): error #6404: This name does not have a type, and must have an explicit type.   [INDSET]
 indset = open_dataset(input_file)
-^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(68): error #6404: This name does not have a type, and must have an explicit type.   [OPEN_DATASET]
 indset = open_dataset(input_file)
----------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(71): error #6404: This name does not have a type, and must have an explicit type.   [NCDIM]
 ncdim = get_dim(indset, 'grid_xt'); i_input = ncdim%len
-^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(71): error #6404: This name does not have a type, and must have an explicit type.   [GET_DIM]
 ncdim = get_dim(indset, 'grid_xt'); i_input = ncdim%len
---------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(71): error #6460: This is not a field name that is defined in the encompassing structure.   [LEN]
 ncdim = get_dim(indset, 'grid_xt'); i_input = ncdim%len
-----------------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(74): error #6404: This name does not have a type, and must have an explicit type.   [GET_IDATE_FROM_TIME_UNITS]
 idate = get_idate_from_time_units(indset)
---------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(174): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'dzdt', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(192): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'rwmr', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(209): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'icmr', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(226): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'snmr', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(243): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'grle', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(261): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
    call read_vardata(indset, 'cld_amt', work3d, errcode=iret)
-------------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(280): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'dpres', work3d, errcode=iret)
--------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(286): error #6632: Keyword arguments are invalid without an explicit interface.   [ERRCODE]
 call read_vardata(indset, 'delz', work3d, errcode=iret)
-------------------------------------------^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(324): error #6404: This name does not have a type, and must have an explicit type.   [REFDSET]
 refdset = open_dataset(ref_file)
-^
/scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90(324): error #6404: This name does not have a type, and must have an explicit type.   [OPEN_DATASET]
 refdset = open_dataset(ref_file)
-----------^
compilation aborted for /scratch1/NCEPDEV/nems/Brian.Curtis/git/BrianCurtis-NOAA/global-workflow/sorc/enkf_chgres_recenter_nc.fd/input_data.f90 (code 1)
make[2]: *** [enkf_chgres_recenter_nc.fd/CMakeFiles/enkf_chgres_recenter_nc.x.dir/input_data.f90.o] Error 1
make[1]: *** [enkf_chgres_recenter_nc.fd/CMakeFiles/enkf_chgres_recenter_nc.x.dir/all] Error 2
```
I'm not sure if you want to get that fixed before merging, as reg2grb2 build succeeds.

WalterKolczynski-NOAA commented:
Are there no changes needed to build_reg2grb2.sh or reg2grb2.fd/Makefile to use cmake?

BrianCurtis-NOAA committed:
Merge branch 'feature/coupled-crow' of https://github.com/NOAA-EMC/global-workflow into feature/reg2grb2_cmake

BrianCurtis-NOAA committed:
Do not need build script nor Makefile. Don't need nemsiogfs dep

BrianCurtis-NOAA commented:
@WalterKolczynski-NOAA OK so the build issue above was actually an issue with hpc-stack that I got taken care of (at least on hera). Make sure that the ncio build from hpc-stack has a module_ncio.mod and not module_fv3gfs_ncio.mod. The script and the makefile are no longer needed. The build completes on Hera.

WalterKolczynski-NOAA committed:
Remove ncio hack

Now that ncio is available in the hpc-stack, removed the workaround
that cloned the ncio repository.

Refs: #268

WalterKolczynski-NOAA committed:
Fix order of build workflow utils

Reordered the build workflow utils script so that machine-setup, which
determines the target machine, is called before $target is used to
determine the appropriate modulefile.

Also added error checking so the script exits immediately if the
modulefile is missing instead of trying to build.

Refs: #268

WalterKolczynski-NOAA committed:
Add missing modules for workflow utils

WGRIB2 and landsfcutil are needed to build workflow utils but were
not included in the modulefiles.

Refs: #268

WalterKolczynski-NOAA committed:
Add reg2grb2 to the link list

With reg2grb2 added to the workflow utils being built with cmake, the
executable needed to be added to the list of executables linked by the
linker script.

Refs: #268

WalterKolczynski-NOAA committed:
Remove separate reg2grb2 from build all

Now that reg2grb2 is part of the workflow_utils cmake build, the old
build script is no longer available and a separate entry is no longer
needed for build_all.

Refs: #268

WalterKolczynski-NOAA commented:
@BrianCurtis-NOAA Take a look at the additional changes I made. System now builds using build_all and link_fv3gfs creates the appropriate link to the executable in the exec directory.

BrianCurtis-NOAA commented:
Looks good to me! Thanks.

aerorahul commented on sorc/build_workflow_utils.sh:
```diff
@@ -5,35 +5,19 @@ set -eux
 
 # Adapt for global-workflow structure.
-target=${target:-"NULL"}
+source ${UTILS_DIR}/machine-setup.sh > /dev/null 2>&1
```
why is this required?
which mahcine is requiring this to be sourced that supports the global workflow?

WalterKolczynski-NOAA replied:
When I tried running build_workflow_utils stand-alone, $target isn't defined until it is set by machine-setup, so it was using NULL, then quietly not loading a modulefile because there wasn't one for NULL, then cmake would fail.

aerorahul commented on sorc/build_workflow_utils.sh:
```diff
@@ -5,35 +5,19 @@ set -eux
   source $modulefile
   module list
   set -x
+else
+  echo "FATAL: modulefile $modulefile not found!"
+  exit 1
```
This assumes that on every machine there has to be a modulefile.
If the software stack is in path, a modulefile is not required.
Which machine is this failing on?

WalterKolczynski-NOAA replied:
The need for this came to me based on the above comment. In the context of global-workflow, I don't know that manually loading a module beforehand is a use case to support, and I'd prefer the error to specify the actual problem rather than cmake just failing.

aerorahul commented on sorc/reg2grb2.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+add_executable(${exe_name} ${fortran_src})
+target_link_libraries(
+  ${exe_name}
+  landsfcutil::landsfcutil_d
+  ip::ip_d
+  sp::sp_d
+  bacio::bacio_4
+  w3nco::w3nco_4
```
has anyone verified that all of these are actually used in this program?

BrianCurtis-NOAA replied:
I went through the makefile and compared which ones were being linked. I didn't go through the code line by line.

WalterKolczynski-NOAA replied:
That's a question for @BrianCurtis-NOAA, but landsfcutil is a new dependency for this PR, so I assume yes for that at least.

BrianCurtis-NOAA replied:
@aerorahul the old build_reg2grb2.sh has landsfcutil in a LIBS export which was then used in the makefile that was used to build reg2grb2. I don't see any subroutines that seem to be using any landsfcutil subroutines, but if needed I will spend more time on going line by line to see if i missed something. Let me know.

aerorahul reviewed (commented):
please see comments.

WalterKolczynski-NOAA commented:
@aerorahul Unless you want me to hold this, I'd like to get it merged.

aerorahul commented:
> @aerorahul Unless you want me to hold this, I'd like to get it merged.

Please go ahead.
